pkg/rpc/server/server.go

```diff
@@ -1974,55 +1997,61 @@ func (s *Server) runScriptInVM(t trigger.Type, script []byte, contractScriptHash
 	if err != nil {
 		faultException = err.Error()
 	}
-	var (
-		registerSession result.OnNewSession
-		params          *result.InvocationParams
-	)
+	var registerSession func(sessionID string, item stackitem.Item, id int, finalize func()) uuid.UUID
 	if s.config.SessionEnabled {
-		registerSession = s.registerSession
-		if s.config.SessionBackedByMPT {
-			params = &result.InvocationParams{
-				Trigger:            t,
-				Script:             script,
-				ContractScriptHash: contractScriptHash,
-				Transaction:        tx,
-				NextBlockHeight:    ic.Block.Index,
+		registerSession = func(sessionID string, item stackitem.Item, stackIndex int, finalize func()) uuid.UUID {
+			iterID := uuid.New()
+			s.sessionsLock.Lock()
+			sess, ok := s.sessions[sessionID]
+			if !ok {
+				timer := time.AfterFunc(time.Second*time.Duration(s.config.SessionExpirationTime), func() {
+					s.sessionsLock.Lock()
+					defer s.sessionsLock.Unlock()
+					if len(s.sessions) == 0 {
+						return
+					}
+					sess, ok := s.sessions[sessionID]
+					if !ok {
+						return
+					}
+					sess.iteratorsLock.Lock()
+					if sess.finalize != nil {
+						sess.finalize()
+					}
+					delete(s.sessions, sessionID)
+					sess.iteratorsLock.Unlock()
+				})
+				sess = &session{
+					finalize: finalize,
+					timer:    timer,
+				}
+				if s.config.SessionBackedByMPT {
+					sess.params = &invocationParams{
+						Trigger:            t,
+						Script:             script,
+						ContractScriptHash: contractScriptHash,
+						Transaction:        tx,
+						NextBlockHeight:    ic.Block.Index,
+					}
+					// Call finalizer manually if MPT-based iterator sessions are enabled. If disabled, then register finalizator.
+					if finalize != nil {
```

Comment:
And this looks like finalizing right after the first iterator registration, while we can have a number of them.

Reply:
OK